### ISSUE TITLE
Include version information even in builds from source

### ISSUE THREAD

**Issue by colemickens** [Masterminds/glide#268]
Hi!

Small feature request. I updated today to get the new 0.9.0 goodness. To be fair, I'm not being very dilligent and am simply running `go get -u github.com/Masterminds/glide`, but it would be nice if `glide --version` would output something other than `glide version dev`.

For example, if my checkout is on git commit `XYZ` and the last released version is `0.9.0`, then maybe it says `glide version 0.9.0-XYZ`. Or if I'm on commit `XYZ` and have local changes that aren't committed, it says `glide version 0.9.0-XYZ-dirty`.

**Comment by mattfarina** [comment]
This would be useful. Not yet sure of the best to handle this so I'm open to suggestions.

**Comment by mhoglan** [comment]
Here is a repo I put together with a pared down example of how I do this in our projects.    

https://github.com/mhoglan/golang_version_injection

The end result is your golang binary will be able to output something like this:

```
root@3e32bcca8e8e:/go/src/github.com/mhoglan/golang_version_injection# /go/bin/golang_version_injection version
{
  "version_info": {
    "branch": "master",
    "build_date": "20160218.164912",
    "build_label": "projectname-v0.0.1-0-ga",
    "commits": "0",
    "dirty": "false",
    "git_describe": "v0.0.1-0-g0034474",
    "git_sha1": "g0034474",
    "label": "ga",
    "version": "v0.0.1"
  }
}
```

Which tells me everything I need to know about the workspace it was generated from.  It also supports injecting runtime based values with ENV var injection which we used to have our build servers inject host information etc...   

Got pretty in depth in this subject as you can see :)

You are welcome to take this and use it.    I am working on eventually opening up the whole example this was pulled from which we use for bootstrapping golang projects.

**Comment by technosophos** [comment]
Does ` go get` not run the Makefile's `build` target? For some reason I thought Go had some baked-in logic for using Makefiles if they were present (instead of just running `go build`).

**Comment by mattfarina** [comment]
Go doesn't do anything special for Make that I can find. There are 3 phases outlined in [`runGet`](https://github.com/golang/go/blob/bc8458ab02878ae64af860f1cade78b6fa97e994/src/cmd/go/get.go#L85). The docs note that [the build tool can only build Go source files and they did not do anything with Makefiles](https://github.com/golang/go/blob/f78a4c84ac8ed44aaf331989aa32e40081fd8f13/doc/articles/go_command.html#L60).

**Comment by mhoglan** [comment]
I did originally look to find some native `go` way to do these kind of things.  The `go generate` that came about in 1.4 was the closest to having any ability for hooking into the `go` tools.   

Ended up choosing `Makefile` because its pretty universal across distribution types and environments without introducing a ton of dependencies on the systems (mainly the dev systems).  Wasn't my ideal choice because of the level of complex logic we needed to get into all of our goals.   For instance the overall `Makefile` I took that snippet out of actually goes off and makes binaries, tarballs, RPMs, containers, etc... It also wraps `docker` commands to perform all of that work inside build containers.  We created it as a way to bootstrap our internal golang projects and have a way to produce the multiple artifacts we need.   

Now for just version information injection, you could achieve that with just the `go generate` tie in.  The `go generate` command that I have in that repo runs the `scripts/includetxt.go` which converts a text file to a golang source file constant.  That text file is just the output of the `generate_version_info.sh` script.  The `go generate` script could have also ran the `generate_version_info.sh` or related `git` commands to generate that constant value.    

Ill see if I can put together just a `go generate` and `go build` example.   

Other note... I know I show how you can inject symbols with the `ldflags` on `go build`, we don't ever use that.  We just did it and showed it as an example.  Its actually a pain to inject json that way.  Originally we were just inject a single label, but it grew into that we would prefer to have a complete dictionary of values.

**Comment by technosophos** [comment]
Just for some background, we've used the linker injection method for quite a while:

https://github.com/Masterminds/glide/blob/master/Makefile#L5

It works really well as long as you build with `make`, and not `go get`. The problem with `go get` is that there is just no convenient way to have it inject the information at build time.

We _could_ generate  a version.go file at tagging time (e.g. `go generate && git commit version.go && git tag $VERSION`). It won't be as precise as the other method, but I suppose it's better than `dev`.

**Comment by hasmichaelh** [comment]
I don't think you want to have to keep a file in the source like a `version.go` that has to be bumped each time.  It will constantly be out of alignment as things occur.  Especially on branches and development work.   Relying on just the tagging annotation and the git workspace information like commits ahead, branch etc... is good.  

Also a build is more than just the `git` information, it is also the environment that it came from.  That was our reason for injecting it all at build time and generating the version information then and not checking it in.  I wanted to be able to check any binary on a system and know if it was an official blessed build or did someone build it locally and transfer it in.

Are you bringing up the `go get` as you want to be able to have a solution that will allow someone to download the source via `go get` which will automatically build and install it?  But I don't believe it will automatically do `go generate`.

**Comment by mattfarina** [comment]
The problem with Makefiles is Windows. Go and Glide support windows.

**Comment by sdboyer** [comment]
> Does go get not run the Makefile's build target? For some reason I thought Go had some baked-in logic for using Makefiles if they were present (instead of just running go build).

Hopefully we can end up shoehorning this one in via manifest...over time :)

**Comment by mattfarina** [comment]
@sdboyer you can't shoehorn in Makefiles because you don't have make in every system Go is available in. Or, did you mean something else?

**Comment by sdboyer** [comment]
@mattfarina - i was referring to a different idea, something analogous to [rust/crates' build script](http://doc.crates.io/build-script.html) system.

obviously there's a *lot* to be figured out in introducing such a feature; i don't reference it lightly.
